WalletWasabi/WabiSabi/Backend/DoSPrevention/Warden.cs

```diff
@@ -2,42 +2,53 @@
 using System.IO;
 using System.Linq;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
-using WalletWasabi.Bases;
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
 using WalletWasabi.Helpers;
 using WalletWasabi.Logging;
+using WalletWasabi.WabiSabi.Backend.Rounds.CoinJoinStorage;
 
 namespace WalletWasabi.WabiSabi.Backend.DoSPrevention;
 
-/// <summary>
-/// Serializes and releases the prison population periodically.
-/// </summary>
-public class Warden : PeriodicRunner
+public class Warden : BackgroundService
 {
-	/// <param name="period">How often to serialize and release inmates.</param>
-	public Warden(TimeSpan period, string prisonFilePath, WabiSabiConfig config) : base(period)
+	public Warden(string prisonFilePath, ICoinJoinIdStore coinjoinIdStore, WabiSabiConfig config)
 	{
 		PrisonFilePath = prisonFilePath;
-		Config = config;
-		Prison = DeserializePrison(PrisonFilePath);
-		LastKnownChange = Prison.ChangeId;
+		OffendersToSaveChannel = Channel.CreateUnbounded<Offender>();
+
+		var dosConfig = new DoSConfiguration(
+			SeverityInBitcoinsPerHour: config.DoSSeverity.ToDecimal(MoneyUnit.BTC),
+			MinTimeForFailedToVerify: config.DoSMinTimeForFailedToVerify,
+			MinTimeForCheating: config.DoSMinTimeForCheating,
+			PenaltyFactorForDisruptingConfirmation: (decimal) config.DoSPenaltyFactorForDisruptingConfirmation,
+			PenaltyFactorForDisruptingSigning: (decimal) config.DoSPenaltyFactorForDisruptingSigning,
+			PenaltyFactorForDisruptingByDoubleSpending: (decimal) config.DoSPenaltyFactorForDisruptingByDoubleSpending,
+			MinimumTimeInPrison: config.DoSMinimumTimeInPrison);
+		Prison = DeserializePrison(PrisonFilePath, dosConfig, coinjoinIdStore, OffendersToSaveChannel.Writer);
 	}
 
 	public Prison Prison { get; }
-	public Guid LastKnownChange { get; private set; }
 
 	public string PrisonFilePath { get; }
-	public WabiSabiConfig Config { get; }
 
-	private static Prison DeserializePrison(string prisonFilePath)
+	private Channel<Offender> OffendersToSaveChannel { get; }
+
+	private static Prison DeserializePrison(
+		string prisonFilePath,
```

Comment:
true. i will take a look at the whole thing because this typically means that our design makes no sense. Ofc nothing happens if this remains as it is because it is not a huge smell as the ones you are fighting in other places of the codebase.